### PR TITLE
Allow multiple overmap terrains in start location

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Infrastructure "Allow multiple overmap terrains in start location"```

#### Purpose of change

Allow multiple overmap terrains in start location without cluttering UI for each distinct terrain.

#### Describe the solution

Code was changed to allow multiple terrains be defined for start locations, they all would occupy a single line in UI and be selected randomly on game start. From players perspective this would allow players to start in any house type instead of only 6 predefined houses. Thanks to overmap match type support it is still possible to create a mod that allow starting in specific terrains.

***Screenshot***:

image

***Syntax:***

```JSON
  {
    "type": "start_location",
    "id": "sloc_police",
    "name": "Police Station",
    "terrain": [ { "om_terrain": "police", "om_terrain_match_type": "PREFIX" } ]
  },
  {
    "type": "start_location",
    "id": "sloc_restaraunt_foodplace_break_room",
    "name": "Foodplace Break Room",
    "terrain": [ "s_restaurant_foodplace_roof" ]
  },
  {
    "type": "start_location",
    "id": "sloc_school",
    "name": "School",
    "terrain": [ { "om_terrain": "school", "om_terrain_match_type": "CONTAINS" } ]
  },
  {
    "type": "start_location",
    "id": "sloc_campground",
    "name": "Campground",
    "terrain": [ "campground_1a", "campground_1b", "campground_2a", "campground_2b" ]
  },
```

I've also updated names of some start locations for consistency.

#### Testing

Try various scenarios and see that player character is put to correct start locations.

#### Additional context

1. I've updated start locations and scenarios in all in-repo mods, but it could possibly break third party mods that add/expand start locations. These mods need to use new syntax:

* replace `target` string node with `terrain` array node (e.g. `"target": "s_bunker_shop_g"` -> `"terrain": [ "s_bunker_shop_g" ]` );
* optionally utilize overmap terrain match type nodes (e.g. `"terrain": [ { "om_terrain": "house", "om_terrain_match_type": "PREFIX" } ]` to reduce UI cluttering).

2. We would probably like to cleanup and expand start locations using new syntax in later PRs (as current PR could possibly allow spawning players in undesired location - e.g. house roof).

### CODE REVIEW

File: src/start_location.cpp [hunk 1]
```diff
@@ -20,78 +19,70 @@
...
+    ot_match_type ter_match_type = ot_match_type::type;
+    for( const JsonValue entry : jo.get_array( "terrain" ) ) {
+        if( entry.test_string() ) {
+            ter = entry.get_string();
+        } else {
+            JsonObject jo = entry.get_object();
+            ter = jo.get_string( "om_terrain" );
+            ter_match_type = jo.get_enum_value<ot_match_type>( "om_terrain_match_type", ter_match_type );
+        }
+        _omt_types.push_back( std::make_pair( ter, ter_match_type ) );
```
Comment (BevapDin) [inline src/start_location.cpp]:
```suggestion
        _omt_types.emplace_back( ter, ter_match_type );
```
It's a bit shorter.

File: src/start_location.h [hunk 2]
```diff
@@ -7,31 +7,28 @@
 #include <vector>
 #include <string>
 
-#include "string_id.h"
+#include "generic_factory.h"
```
Comment (BevapDin) [inline src/start_location.h]: This seems to be unused in the header. It should only be in the cpp file.

**Comment by AMurkin** [comment]
Could the first option of that list be a `Random location`? (Looking at the #38110)

**Comment by ZhilkinSerg** [comment]
> 
> 
> Could the first option of that list be a `Random location`? (Looking at the #38110)

Sure. It is unrelated though and can be done in a separate PR.

File: data/json/start_locations.json [hunk 3]
```diff
@@ -1,437 +1,408 @@
...
-    "ident": "mil_base_2g",
+    "id": "sloc_military_base_warehouse",
     "name": "Military Base Warehouse",
-    "target": "mil_base_2g"
+    "terrain": [ "mil_base_2g" ]
   },
   {
     "type": "start_location",
-    "ident": "p_resort_2ss",
+    "ident": "sloc_private_resort",
```
Comment (ZhilkinSerg) [inline data/json/start_locations.json:404]:
```suggestion
    "id": "sloc_private_resort",
```

**Comment by ZhilkinSerg** [comment]
> 
> 
> Could the first option of that list be a `Random location`? (Looking at the #38110)

It was done in #38544.

**Comment by ZhilkinSerg** [comment]
There is a strange bug I cannot grok by myself:

When you select Evacuee scenario (only it as far as I see) you cannot select a second entry in starting location selection menu.

**Pull request merged**